### PR TITLE
Allow repositories to manage sidebars and other content

### PR DESCRIPTION
**Pull request by schpet**
todo

- [x] show validation errors when a manifest is not valid?
- [x] add a route with JSON schema 
- [x] cleanup code
- [x] establish how cached data will be updated
- [x] explain what content is controlled by the manifest in this PR description
- [x] decide on a good filename for the manifest configuration file, currently it is `<Docs Dir>/flow-docs.json`
- [ ] create and collaborate on documentation
- [ ] make issue to convert the document caching to use the github webhooks, and deprecate the refresh endpoint
- [x] make issue to support loading yaml (done: https://github.com/onflow/next-docs-v1/issues/605)

example manifest:
https://github.com/onflow/mock-developer-doc/blob/json-manifest-valid/docs/

a manifest is able to control:

- display name (e.g. Cadence)
- sidebars
- redirects

example of an external repo loading it's own docs-config.json:

Screen Shot 2022-08-10 at 5 44 42 PM

example of situations where a repo has a broken flow-docs.json (syntax errors, schema errors, etc): 

Screen Shot 2022-08-17 at 4 49 44 PM

### CODE REVIEW

**Comment by schpet** [comment]
@joekrill this is a rough cut, but interested in your take on this

File: app/cms/github-webhook.server.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    ...commit.removed,
+    ...commit.modified,
+  ])
+
+  const keysToInvalidate: Set<string> = new Set()
+
+  const manifestPath = posix.join(
+    docCollection.source.rootPath,
+    JSON_MANIFEST_FILENAME
+  )
```
Comment (schpet) [inline app/cms/github-webhook.server.ts:49]: for the purposes of this PR, i'm going to hold off on invalidating the document cache for md/mdx files – that can happen in a future PR, ideally after the webhooks on the github side are all setup

File: app/cms/github.server.ts [hunk 2]
```diff
@@ -325,7 +325,7 @@ export async function downloadFileByPath(
   const { data } = await octokit.repos.getContent({
     owner: source.owner,
     repo: source.name,
-    headers: { branch: source.branch },
+    ref: source.branch,
```
Comment (schpet) [inline app/cms/github.server.ts:328]: @joekrill can you please sanity check that this is correct? i don't know where the headers thing came from

File: app/root.tsx [hunk 3]
```diff
@@ -66,7 +66,7 @@ export const links: LinksFunction = () => {
 export const meta: MetaFunction = ({ data, location }) => ({
   ...getSocialMetas({
     title: getMetaTitle(),
-    url: data.url,
+    url: data?.url ?? "",
```
Comment (schpet) [inline app/root.tsx:69]: fyi that `data` will be undefined in meta functions when errors happen, throwing in a meta function will mess up error catching

File: app/cms/github.server.ts (hunk 2 above)
Comment (joekrill) [reply app/cms/github.server.ts:328]: I could swear I found this documented somewhere and I remember making this change. But I'm looking for it again and I can't find it anywhere. So I must have misread the docs or something.

**Review by joekrill — APPROVED** [review]

File: app/constants/doc-collection-manifest.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+import { DocCollectionSource } from "./doc-collections.server"
+
+export const JSON_MANIFEST_FILENAME = "flow-docs.json"
```
Comment (schpet) [inline app/constants/doc-collection-manifest.ts:3]: @10thfloor are you good with `docs/flow-docs.json` as the path for external repos to set these values?

**Review by 10thfloor — APPROVED** [review]

**Comment by schpet** [comment]
https://github.com/onflow/next-docs-v1/issues/590

**Pull request merged**